File: frontend/lib/review/models/review.dart
class Review {
  final int id;
  final int rating;
  final String? comment;
  final int userId;
  final int busId;
  final int ticketId;
  final DateTime createdAt;
  final String? userName;
  final String? userAvatar;
  final List<String> images;
  final String? reply;
  final DateTime? repliedAt;

  Review({
    required this.id,
    required this.rating,
    this.comment,
    required this.userId,
    required this.busId,
    required this.ticketId,
    required this.createdAt,
    this.userName,
    this.userAvatar,
    this.images = const [],
    this.reply,
    this.repliedAt,
  });

  factory Review.fromJson(Map<String, dynamic> json) {
    return Review(
      id: json['id'],
      rating: json['rating'],
      comment: json['comment'],
      userId: json['userId'],
      busId: json['busId'],
      ticketId: json['ticketId'],
      createdAt: DateTime.parse(json['createdAt']),
      userName: json['user']?['name'],
      userAvatar: json['user']?['avatar'],
      images: (json['images'] as List<dynamic>?)?.map((e) => e.toString()).toList() ?? [],
      reply: json['reply'],
      repliedAt: json['repliedAt'] != null ? DateTime.parse(json['repliedAt']) : null,
    );
  }
}

File: web/src/lib/api/review.ts
import api from '@/lib/api';
import axios from 'axios';

export interface Review {
    id: number;
    ticketId: number;
    userId: number;
    busId: number;
    rating: number;
    comment: string;
    images?: string[];
    createdAt: string;
    updatedAt: string;
    user?: {
        fullName: string;
        avatar?: string;
    };
}

export interface CreateReviewDto {
    ticketId: number;
    rating: number;
    comment?: string;
    images?: string[];
}

export const reviewApi = {
    // Create a new review
    create: async (data: CreateReviewDto): Promise<Review> => {
        const response = await api.post('/reviews', data);
        return response.data;
    },

    // Get reviews by current user
    getMyReviews: async (): Promise<Review[]> => {
        const response = await api.get('/reviews/my-reviews');
        return response.data;
    },

    // Get unreviewed tickets for current user
    getUnreviewedTickets: async (): Promise<any[]> => {
        const response = await api.get('/reviews/unreviewed');
        return response.data;
    },

    // Get reviews by bus ID
    getByBusId: async (busId: number): Promise<Review[]> => {
        const response = await api.get(`/reviews/bus/${busId}`);
        return response.data;
    },

    // Get review stats by bus ID
    getStats: async (busId: number): Promise<{ avgRating: number; totalReviews: number }> => {
        const response = await api.get(`/reviews/stats/${busId}`);
        return response.data;
    },

    // Upload review image
    uploadImage: async (file: File): Promise<string> => {
        const formData = new FormData();
        formData.append('file', file);

        // Use the configured 'api' instance so the Authorization interceptor attaches the token
        const response = await api.post<{ url: string }>('/upload/review', formData, {
            headers: {
                'Content-Type': 'multipart/form-data',
            }
        });
        return response.data.url;
    }
};
